modules/navier_stokes/src/executioners/SIMPLE.C

```diff
@@ -0,0 +1,1106 @@
+      ns_residuals[momentum_residual.size()] = solvePressureCorrector();
+
+      // Compute the face velocity which is used in the advection terms
+      _rc_uo->computeFaceVelocity();
+
+      // Relax the pressure update for the next momentum predictor
+      relaxSolutionUpdate(_pressure_system, _pressure_variable_relaxation);
+
+      // Reconstruct the cell velocity as well to accelerate convergence
+      _rc_uo->computeCellVelocity();
+
+      // If we have an energy equation, solve it here. We assume the material properties in the
+      // Navier-Stokes equations depend on temperature, therefore we can not solve for temperature
+      // outside of the velocity-pressure loop
+      if (_has_energy_system)
+      {
+        // We set the preconditioner/controllable parameters through petsc options. Linear
+        // tolerances will be overridden within the solver.
+        Moose::PetscSupport::petscSetOptions(_energy_petsc_options, solver_params);
+        ns_residuals[momentum_residual.size() + 1] =
+            solveAdvectedSystem(_energy_sys_number,
+                                *_energy_system,
+                                _energy_equation_relaxation,
+                                _energy_ls_control,
+                                _energy_l_abs_tol);
+
+        if (_has_solid_energy_system)
+        {
+          // We set the preconditioner/controllable parameters through petsc options. Linear
+          // tolerances will be overridden within the solver.
+          Moose::PetscSupport::petscSetOptions(_energy_petsc_options, solver_params);
+          ns_residuals[momentum_residual.size() + 2] = solveSolidEnergySystem();
+        }
+      }
+
+      _console << "Iteration " << iteration_counter << " Initial residual norms:" << std::endl;
+      for (auto system_i : index_range(_momentum_systems))
+        _console << " Momentum equation:"
+                 << (_momentum_systems.size() > 1
+                         ? std::string(" Component ") + std::to_string(system_i + 1) +
+                               std::string(" ")
+                         : std::string(" "))
+                 << COLOR_GREEN << ns_residuals[system_i] << COLOR_DEFAULT << std::endl;
+      _console << " Pressure equation: " << COLOR_GREEN << ns_residuals[momentum_residual.size()]
+               << COLOR_DEFAULT << std::endl;
+      if (_has_energy_system)
+      {
+        _console << " Energy equation: " << COLOR_GREEN
+                 << ns_residuals[momentum_residual.size() + 1] << COLOR_DEFAULT << std::endl;
+        if (_has_solid_energy_system)
+          _console << " Solid energy equation: " << COLOR_GREEN
+                   << ns_residuals[momentum_residual.size() + 2] << COLOR_DEFAULT << std::endl;
+      }
+    }
+
+    // Now we solve for the passive scalar equations, they should not influence the solution of the
+    // system above. The reason why we need more than one iterations is due to the matrix relaxation
```

Comment:
```suggestion
    // system above. The reason why we need more than one iteration is due to the matrix relaxation
```

Reply:
I still dont get why

velocities should be set after the p-v iteration
then we use those velocities once to get the scalars which dont depend on each other either (so only need 1 solve to get them)

Reply:
I thought they may have coupled sources

Reply:
Oh yes and the matrix relaxation. In purely advective problem I have seen issues with scalar transport without relaxation. Well, not necessaily with upwind discretization but with linear and so on. Those can get nonbounded.